test/control_plane_status_test.go

```diff
@@ -0,0 +1,293 @@
+
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	csErr := make(chan error, 1)
+	state := iresolver.SetConfigSelector(resolver.State{
+		Addresses:     []resolver.Address{{Addr: ss.Address}},
+		ServiceConfig: parseServiceConfig(t, ss.R, "{}"),
+	}, funcConfigSelector{
+		f: func(i iresolver.RPCInfo) (*iresolver.RPCConfig, error) {
+			return nil, <-csErr
+		},
+	})
+	ss.R.UpdateState(state) // Blocks until config selector is applied
+
+	testCases := []struct {
+		name  string
+		csErr error
+		want  error
+	}{{
+		name:  "legal status code",
+		csErr: status.Errorf(codes.Unavailable, "this error is fine"),
+		want:  status.Errorf(codes.Unavailable, "this error is fine"),
+	}, {
+		name:  "illegal status code",
+		csErr: status.Errorf(codes.NotFound, "this error is bad"),
+		want:  status.Errorf(codes.Internal, "this error is bad"),
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// In case the channel is full due to a previous iteration failure,
+			// do not block.
+			select {
+			case csErr <- tc.csErr:
+			default:
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			if _, err := ss.Client.EmptyCall(ctx, &testpb.Empty{}); status.Code(err) != status.Code(tc.want) || !strings.Contains(err.Error(), status.Convert(tc.want).Message()) {
+				t.Fatalf("client.EmptyCall(_, _) = _, %v; want _, %v", err, tc.want)
+			}
+		})
+	}
+}
+
+type lbBuilderWrapper struct {
+	builder balancer.Builder // real Builder
+	name    string
+	picker  func(balancer.PickInfo) error
+}
+
+func (l *lbBuilderWrapper) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return l.builder.Build(&lbCCWrapper{ClientConn: cc, picker: l.picker}, opts)
+}
+
+func (l *lbBuilderWrapper) Name() string {
+	return l.name
+}
+
+type lbCCWrapper struct {
+	balancer.ClientConn // real ClientConn
+	picker              func(balancer.PickInfo) error
+}
+
+func (l *lbCCWrapper) UpdateState(s balancer.State) {
+	s.Picker = &lbPickerWrapper{picker: l.picker, Picker: s.Picker}
+	l.ClientConn.UpdateState(s)
+}
+
+type lbPickerWrapper struct {
+	balancer.Picker // real Picker
+	picker          func(balancer.PickInfo) error
+}
+
+func (lp *lbPickerWrapper) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	if err := lp.picker(info); err != nil {
+		return balancer.PickResult{}, err
+	}
+	return lp.Picker.Pick(info)
+}
+
+func (s) TestPickerStatusCodes(t *testing.T) {
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			return &testpb.Empty{}, nil
+		},
+	}
+
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	pickerErr := make(chan error, 1)
+	balancer.Register(&lbBuilderWrapper{
+		builder: balancer.Get("round_robin"),
```

Comment:
Why do we need a real balancer inside this test balancer? Can't we use the stub balancer and override the `UpdateClientConnState` to return an error picker (with an error configured through the test table)?

And similar to the comment on the earlier test, could you please move all setup inside the `t.Run()` so that the subTests can be independent of each other.

Reply:
> Why do we need a real balancer inside this test balancer? Can't we use the stub balancer and override the UpdateClientConnState to return an error picker (with an error configured through the test table)?

Yes, that's much simpler, thanks.

> move all setup inside the t.Run()

Done; simplified.